lib/trie.js

```diff
@@ -0,0 +1,57 @@
+module.exports = Trie;
+
+function Trie() {
+  this.methods = {}
+}
+
+//insert new nodes to Trie
+Trie.prototype.add = function (path, verb, handler) {
+  var node = this;
+  var parts = getPathParts(path);
+
+  if (parts.length) {
+    for (var i = 0; i < parts.length; i++) {
+      if (!node || !(parts[i] in node)) {
+        node[parts[i]] = new Trie();
+      }
+      if (i + 1 === parts.length) {
+        node[parts[i]].methods = node[parts[i]].methods || {};
+        node[parts[i]].methods[verb] = handler;
+      }
+      node = node[parts[i]];
+    }
+  }
+  else { // if we would ever encounter this situation? 
+    node.methods = node.methods || {};
+    node.methods[verb] = handler;
```

Comment:
Ah, of course, you are right. Your proposal to keep all handlers in an array makes perfect sense.